From 8a29f081883b69f76af45754cd1a5a1b5ff63cdf Mon Sep 17 00:00:00 2001
From: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
Date: Tue, 15 Nov 2022 03:04:22 -0500
Subject: ath12k: Sanitize peer list before vdev delete in STA

When doing wifi restart mutiple times with STA, sometimes
mac80211 call unassign_vif_chanctx while STA destroy is not
called.

If vdev delete is sent with any active peers to FW, FW is
expected to assert. Hence before sending vdev delete in case of STA,
check for peers and delete it followed by vdev delete

Signed-off-by: Ramya Gnanasekar <quic_rgnanase@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 37 ++++++++++++++++++++-------
 1 file changed, 28 insertions(+), 9 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -11753,6 +11753,8 @@ ath12k_mac_op_unassign_vif_chanctx(struc
 	struct ath12k_base *ab;
 	struct ath12k_vif *ahvif = ath12k_vif_to_ahvif(vif);
 	struct ath12k_link_vif *arvif;
+	struct ath12k_dp_link_peer *peer;
+	struct ath12k_dp *dp;
 	u8 link_id = link_conf->link_id;
 	int ret;
 
@@ -11772,6 +11774,7 @@ ath12k_mac_op_unassign_vif_chanctx(struc
 
 	ar = arvif->ar;
 	ab = ar->ab;
+	dp = ath12k_ab_to_dp(ab);
 
 	if (unlikely(test_bit(ATH12K_FLAG_CRASH_FLUSH, &ar->ab->dev_flags)))
 		return;
@@ -11800,6 +11803,22 @@ ath12k_mac_op_unassign_vif_chanctx(struc
 	}
 	arvif->is_started = false;
 
+	spin_lock_bh(&dp->dp_lock);
+	peer = ath12k_dp_link_peer_find_by_vdev_id_and_addr(dp, arvif->vdev_id, arvif->bssid);
+	if (ahvif->vdev_type == WMI_VDEV_TYPE_STA && peer) {
+		ret = ath12k_peer_delete(ar, arvif->vdev_id, arvif->bssid);
+		if (ret) {
+			ath12k_err(ar->ab,
+				   "failed to delete peer %pM for vdev %d: %d\n",
+				   arvif->bssid, arvif->vdev_id, ret);
+		} else {
+			ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
+				   "mac removed peer %pM  vdev %d\n",
+				   arvif->bssid, arvif->vdev_id);
+		}
+	}
+	spin_unlock_bh(&dp->dp_lock);
+
 	if (ar->scan.arvif == arvif && ar->scan.state == ATH12K_SCAN_RUNNING) {
 		ath12k_scan_abort(ar);
 		ar->scan.arvif = NULL;
